From a49d5417cfcb1343eff8ff001cc56062628e805a Mon Sep 17 00:00:00 2001
From: Artem Butusov <art.sormy@gmail.com>
Date: Tue, 13 Oct 2026 14:22:59 -0400
Subject: [PATCH 2/5] avcodec/v4l2_m2m_dec: send the extradata of codecs
 without a header filter

H.264 and HEVC reach the driver with their parameter sets through
the mp4toannexb filters. MPEG-1/2, MPEG-4, H.263 and VC-1 from a
container carry their sequence header in extradata only, which was
never sent: a VC-1 stream from Matroska decoded nothing. Send it as
its own buffer ahead of the first packet.
---
 libavcodec/v4l2_m2m.h     |  3 +++
 libavcodec/v4l2_m2m_dec.c | 42 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 45 insertions(+)

diff --git a/libavcodec/v4l2_m2m.h b/libavcodec/v4l2_m2m.h
index f8b92f0..c874a87 100644
--- a/libavcodec/v4l2_m2m.h
+++ b/libavcodec/v4l2_m2m.h
@@ -62,6 +62,9 @@ typedef struct V4L2m2mContext {
     int no_pts;
     int64_t nopts_seq;
 
+    /* extradata went out ahead of the first packet */
+    int extradata_sent;
+
     /* Reference to a frame. Only used during encoding */
     AVFrame *frame;
 
diff --git a/libavcodec/v4l2_m2m_dec.c b/libavcodec/v4l2_m2m_dec.c
index ba1fe76..af12486 100644
--- a/libavcodec/v4l2_m2m_dec.c
+++ b/libavcodec/v4l2_m2m_dec.c
@@ -133,6 +133,40 @@ static int v4l2_prepare_decoder(V4L2m2mContext *s)
     return 0;
 }
 
+/* The container's sequence header reaches the driver only as extradata. */
+static int v4l2_needs_extradata(enum AVCodecID id)
+{
+    return id == AV_CODEC_ID_MPEG1VIDEO || id == AV_CODEC_ID_MPEG2VIDEO ||
+           id == AV_CODEC_ID_MPEG4 || id == AV_CODEC_ID_H263 ||
+           id == AV_CODEC_ID_VC1;
+}
+
+static int v4l2_send_extradata(AVCodecContext *avctx, V4L2m2mContext *s)
+{
+    AVPacket *pkt;
+    int ret;
+
+    if (s->extradata_sent || !avctx->extradata_size ||
+        !v4l2_needs_extradata(avctx->codec_id))
+        return 0;
+
+    pkt = av_packet_alloc();
+    if (!pkt)
+        return AVERROR(ENOMEM);
+    ret = av_new_packet(pkt, avctx->extradata_size);
+    if (ret < 0)
+        goto end;
+    memcpy(pkt->data, avctx->extradata, avctx->extradata_size);
+    pkt->pts = s->buf_pkt.pts;
+
+    ret = ff_v4l2_context_enqueue_packet(&s->output, pkt);
+    if (!ret)
+        s->extradata_sent = 1;
+end:
+    av_packet_free(&pkt);
+    return ret;
+}
+
 static int v4l2_receive_frame(AVCodecContext *avctx, AVFrame *frame)
 {
     V4L2m2mContext *s = ((V4L2m2mPriv*)avctx->priv_data)->context;
@@ -153,6 +187,14 @@ static int v4l2_receive_frame(AVCodecContext *avctx, AVFrame *frame)
     if (s->draining)
         goto dequeue;
 
+    if (s->buf_pkt.size) {
+        ret = v4l2_send_extradata(avctx, s);
+        if (ret < 0 && ret != AVERROR(EAGAIN))
+            goto fail;
+        if (ret == AVERROR(EAGAIN))
+            goto dequeue;
+    }
+
     ret = ff_v4l2_context_enqueue_packet(output, &s->buf_pkt);
     if (ret < 0 && ret != AVERROR(EAGAIN))
         goto fail;
-- 
2.55.0

